x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java

```diff
@@ -0,0 +1,298 @@
+                prepareAndCreateEnrichIndex(policyName, policy, listener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        });
+    }
+
+    private Map<String, Object> getMappings(final GetIndexResponse getIndexResponse, final String sourceIndexName) {
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = getIndexResponse.mappings();
+        ImmutableOpenMap<String, MappingMetaData> indexMapping = mappings.get(sourceIndexName);
+        assert indexMapping.keys().size() == 0 : "Expecting only one type per index";
+        MappingMetaData typeMapping = indexMapping.iterator().next().value;
+        return typeMapping.sourceAsMap();
+    }
+
+    private void validateMappings(final GetIndexResponse getIndexResponse, final String policyName, final EnrichPolicy policy,
+                                  final ActionListener<?> listener) {
+        String[] sourceIndices = getIndexResponse.getIndices();
+        logger.debug("Policy [{}]: Validating [{}] source mappings", policyName, sourceIndices);
+        for (String sourceIndex : sourceIndices) {
+            Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndex);
+            Map<?, ?> properties = ((Map<?, ?>) mapping.get("properties"));
+            if (properties == null) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
+                        policyName, sourceIndex, policy.getIndexPattern()));
+            }
+            if (properties.containsKey(policy.getEnrichKey()) == false) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not locate enrich key field [{}] on mapping for index [{}]",
+                        policyName, policy.getEnrichKey(), sourceIndex));
+            }
+            for (String enrichField : policy.getEnrichValues()) {
+                if (properties.containsKey(enrichField) == false) {
+                    listener.onFailure(new ElasticsearchException(
+                        "Enrich policy execution for [{}] failed. Could not locate enrich value field [{}] on mapping for index [{}]",
+                        policyName, enrichField, sourceIndex));
+                }
+            }
+        }
+    }
+
+    private String getEnrichIndexBase(final String policyName) {
+        return ".enrich-" + policyName;
+    }
+
+    private Map<String, Object> resolveEnrichMapping(final EnrichPolicy policy) {
+        // TODO: Modify enrichKeyFieldMapping based on policy type.
+        // Create dynamic mapping templates - one for the enrich key and one for the enrich values
```

Comment:
Is it possible to make this a bit more readable via a single map with indentions ?

..or a `JsonXContent.contentBuilder().startObject()` and `Strings.toString(templateBuilder)`